### PR TITLE
Add CLR ABI document to Book of the Runtime

### PR DESCRIPTION
**Pull request by BruceForstall**

### CODE REVIEW

**Comment by BruceForstall** [comment]
@dotnet/jit-contrib @jkotas PTAL

**Comment by janvorli** [comment]
It would be good to mention in the last section that on Unix, we don't support native varargs. And I would also mention in the last bullet of the last section that the registers used are different from Windows (since you say "see the relevant sections above" and those sections explicitly mention register names like RCX or RDX)

File: Documentation/botr/clr-abi.md [hunk 1]
```diff
@@ -0,0 +1,550 @@
...
+
+# PInvokes
+
+The convention is that any method with an InlinedCallFrame (either an IL stub or a normal method with an inlined pinvoke) saves/restores all non-volatile integer registers in its prolog/epilog respectively. This is done so that the InlinedCallFrame can just contain a return address, a stack pointer and a frame pointer. Then using just those three it can start a full stack walk using the normal RtlVirtualUnwind.
+
+For AMD64, a method with an InlinedCallFrame must use RBP as the frame register.
+
+For ARM and ARM64, we will also always use a frame pointer (R11), but that is due to the frame chaining requirement. I do not know if the VM also requires it or has taken a dependency upon it for PInvokes.
```
Comment (jkotas) [inline Documentation/botr/clr-abi.md]: It does have dependency on it. All of these dependencies show up in `InlinedCallFrame::UpdateRegDisplay`.

Looking at `InlinedCallFrame::UpdateRegDisplay`, there is also a dependency on `REG_SAVED_LOCALLOC_SP` that is not mentioned here.

File: Documentation/botr/clr-abi.md [hunk 2]
```diff
@@ -0,0 +1,550 @@
...
+
+JIT64, and RyuJIT for AMD64 and ARM64, creates something similar to a jump island: a block of code outside the try body that calls the finally and then branches to the final target of the leave/non-local-exit. This jump island is then marked in the EH tables as if it were a cloned finally. The cloned finally clause prevents a Thread.Abort from firing before entering the handler. By having the return address outside of the try body we satisfy the other constraint.
+
+Note that ARM solves this by not using a call (bl) instruction and instead explicitly places a return address in `LR` and then jumps to the finally. We have not yet implemented this for AMD64 because it might mess up the call-return predictor on the CPU. (So far performance data on ARM indicates they don't have an issue).
+
+## ThreadAbortException considerations
+
```
Comment (jkotas) [inline Documentation/botr/clr-abi.md]: Do we want to mention that Thread abort is not supported in .NET Core, but we still care about it for full framework?

Comment (jkotas) [inline Documentation/botr/clr-abi.md]: (Similar for other full framework specific features like managed varargs.)

**Comment by jkotas** [comment]
LGTM

File: Documentation/botr/clr-abi.md [hunk 3]
```diff
@@ -0,0 +1,550 @@
...
+### Example
+```
+call(["this" pointer] [return buffer pointer] [generics context|varargs cookie] [userargs]*)
+```
+
+## AMD64-only: by-value value types
+
+Just like native, AMD64 has implicit-byrefs. Any structure (value type in IL parlance) that is not 1, 2, 4, or 8 bytes in size (i.e., 3, 5, 6, 7, or >= 9 bytes in size) that is declared to be passed by value, is instead passed by reference. For JIT generated code, it follows the native ABI where the passed-in reference is a pointer to a compiler generated temp local on the stack. However, there are some cases within remoting or reflection where apparently stackalloc is too hard, and so they pass in pointers within the GC heap, thus the JITed code must report these implicit byref parameters as interior pointers (BYREFs in JIT parlance), in case the callee is one of these reflection paths. Similarly, all writes must use checked write barriers.
+
```
Comment (mikedn) [inline Documentation/botr/clr-abi.md]: > For JIT generated code, it follows the native ABI where the passed-in reference is a pointer to a compiler generated temp local on the stack.

How set in stone is this convention? In addition to the requirement for GC write barriers this also prevents some compiler optimizations, LICM for example. Passing a byref to the original value and copying the value inside the callee would avoid such issues.

File: Documentation/botr/clr-abi.md [hunk 4]
```diff
@@ -0,0 +1,550 @@
...
+# System V x86_64 support
+
+This section relates mostly to calling conventions on System V systems (such as Ubuntu Linux and Mac OS X).
+The general rules outlined in the System V x86_64 ABI (described at http://www.x86-64.org/documentation/abi.pdf) are followed with a few exceptions, described below:
+
+1. The hidden argument for by-value passed structs is always after the "this" parameter (if there is one). This is a difference with the System V ABI and affects only the internal JIT calling conventions. For PInvoke calls the hidden argument is always the first parameter since there is no "this" parameter in this case.
+2. Managed structs that have no fields are always passed by-value on the stack.
+3. The JIT proactively generates frame register frames (with `RBP` as a frame register) in order to aid the native OS tooling for stack unwinding and the like.
+4. All the other internal VM contracts for PInvoke, EH, and generic support remains in place. Please see the relevant sections above for more details.
+
```
Comment (LLITCHEV) [inline Documentation/botr/clr-abi.md]: One more thing that comes to mind is that structs with explicit layouts are always passed by value on the stack.

File: Documentation/botr/clr-abi.md (hunk 3 above)
Comment (BruceForstall) [inline Documentation/botr/clr-abi.md]: I would say this calling convention is fairly set in stone at this point.

**Pull request merged**